runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -753,6 +754,13 @@ private void maybeEmitWatermark(long watermark) {
       LOG.debug("Emitting watermark {}", watermark);
       currentOutputWatermark = watermark;
       output.emitWatermark(new Watermark(watermark));
+
+      // Check if the final watermark was triggered to perform state cleanup for global window
+      if (keyedStateInternals != null
+          && currentOutputWatermark
```

Comment:
The existing solution checks that we have not already reached the watermark to avoid repeated execution of the cleanup code. Please carry that over.

Also, why check for `keyedStateInternals`? See https://github.com/apache/beam/pull/12733#discussion_r480015675

Reply:
Cleanup will only be performed if there is new state which IMHO is the correct thing to do.

Reply:
Due to all logic now being in keyedStateInternals, we need the null check. The state internals will be null if we do not have keyed input.